src/libsyntax/parse/lexer/mod.rs

```diff
@@ -433,6 +436,104 @@ impl<'a> StringReader<'a> {
         Ok(())
     }
 
+    fn advance_with_rust_lexer(&mut self) -> bool {
```

Comment:
This function is not really pretty, this should improve quite a bit in the final implementation: everything will be parsed by the new lexer, so the `bool` token will go. Additionally, the `StringReader` here will become stateless, so `pos` synchronization wouldn't be necessary.